bundle/run/app.go

```diff
@@ -0,0 +1,258 @@
+		logProgress(ctx, fmt.Sprintf("App is in %s state", createdApp.AppStatus.State))
+	}
+
+	if createdApp.ComputeStatus != nil {
+		logProgress(ctx, fmt.Sprintf("App compute is in %s state", createdApp.ComputeStatus.State))
+	}
+
+	// There could be 2 reasons why the app is not running:
+	// 1. The app is new and was never deployed yet.
+	// 2. The app was stopped (compute not running).
+	// We need to start the app only if the compute is not running.
+	if isAppStopped(createdApp) {
+		err := a.start(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Deploy the app.
+	err = a.deploy(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cmdio.LogString(ctx, fmt.Sprintf("You can access the app at %s", createdApp.Url))
+	return nil, nil
+}
+
+func (a *appRunner) start(ctx context.Context) error {
+	app := a.app
+	b := a.bundle
+	w := b.WorkspaceClient()
+
+	logProgress(ctx, fmt.Sprintf("Starting the app %s", app.Name))
+	wait, err := w.Apps.Start(ctx, apps.StartAppRequest{Name: app.Name})
+	if err != nil {
+		return err
+	}
+
+	startedApp, err := wait.OnProgress(func(p *apps.App) {
+		if p.AppStatus == nil {
+			return
+		}
+		logProgress(ctx, "App is starting...")
+	}).Get()
+
+	if err != nil {
+		return err
+	}
+
+	if startedApp.ActiveDeployment != nil &&
+		startedApp.ActiveDeployment.Status.State == apps.AppDeploymentStateInProgress {
+		logProgress(ctx, "Waiting for the active deployment to complete...")
+		_, err = w.Apps.WaitGetDeploymentAppSucceeded(ctx, app.Name, startedApp.ActiveDeployment.DeploymentId, 20*time.Minute, nil)
+		if err != nil {
+			return err
+		}
+		logProgress(ctx, "Active deployment is completed!")
+	}
+
+	if startedApp.PendingDeployment != nil &&
+		startedApp.PendingDeployment.Status.State == apps.AppDeploymentStateInProgress {
+		logProgress(ctx, "Waiting for the pending deployment to complete...")
+		_, err = w.Apps.WaitGetDeploymentAppSucceeded(ctx, app.Name, startedApp.PendingDeployment.DeploymentId, 20*time.Minute, nil)
+		if err != nil {
+			return err
+		}
+		logProgress(ctx, "Pending deployment is completed!")
+	}
+
+	logProgress(ctx, "App is started!")
+	return nil
+}
+
+func (a *appRunner) deploy(ctx context.Context) error {
+	app := a.app
+	b := a.bundle
+	w := b.WorkspaceClient()
+
+	// If the app has a config, we need to deploy it first.
+	// It means we need to write app.yml file with the content of the config field
+	// to the remote source code path of the app.
+	if app.Config != nil {
+		appPath, err := filepath.Rel(b.Config.Workspace.FilePath, app.SourceCodePath)
+		if err != nil {
+			return fmt.Errorf("failed to get relative path of app source code path: %w", err)
+		}
+
+		buf, err := configToYaml(app)
+		if err != nil {
+			return err
+		}
+
+		// When the app is started, create a new app deployment and wait for it to complete.
+		f, err := a.filerFactory(b)
+		if err != nil {
+			return err
+		}
+
+		err = f.Write(ctx, path.Join(appPath, "app.yml"), buf, filer.OverwriteIfExists)
```

Comment:
What is the trade-off to write this on run vs on deploy?

I can imagine someone running a deploy and then clicking "start" on the app. If it auto-picks the previous source code location, my expectation would be that it is ready to go, and has the up-to-date `app.yml`

Reply:
The way I see it is that this configuration is tight to deployment not an app (different deployments have different configs). Since we do the deployment at run, it makes sense to write this at run as well. Having it as part of run gives you a faster loop: for example, if you need to add a flag or env variable to run you app different you jut change a config and execute `databricks bundle run my_app`.

Reply:
That's true, but it also means you _have_ to call run to make the deployment complete, which is counterintuitive to me. If the app ends up in the stopped state and you go into the UI and click start, it will use the path of the most recent deployment. If this gets changed and deployment is decoupled from the compute, then we'll have to move generating this file into bundle deployment.

Do you think we can swap this around later without issues?

Reply:
Also, it would be great to capture this rationale in a comment somewhere in this file.

Reply:
Giving it another thought I think writing this file on deploy rather than run is a better approach from DABs perspective, it makes the flow the same for all cases (deploy first, then run) and also more future proof: it does not matter if config will be part of create app or create app deployment API, DABs will have the same flow anyway